[PyTorch Edge] bytecode version bump to v5 and enable share constant table

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57888 [PyTorch Edge] bytecode version bump to v5 and enable share constant table**

As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

Commit:
[PyTorch Edge] bytecode version bump to v5 and enable share constant table

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]

iseeyuan commented on caffe2/serialize/versions.h:
```diff
@@ -67,7 +67,11 @@ constexpr uint64_t kProducedFileFormatVersion = 0x3L;
 //  0x4L: (update) Added schema to function tuple. Forward-compatible change.
-constexpr uint64_t kProducedBytecodeVersion = 0x4L;
+//  0x5L: (update) Update tensor storage schema. For bytecode archive, the
+//  root key of tensor storage is updated from {index} to {the_pointer_value_the_tensor.storage}.
```
Add the comment that bytecode is sharing constant tensor files from torchscript, and only serialize extra tensors that are not in the torchscript constant table.

cccclai replied:
Add it in last commit

iseeyuan reviewed (approved):
LGTM. Before it's landed, please make sure the service side is aware of this bump, and also they are already using the backport function.

raziel commented on caffe2/serialize/versions.h:
```diff
@@ -67,7 +67,11 @@ constexpr uint64_t kProducedFileFormatVersion = 0x3L;
 //  0x3L: (Comment missing)
 //  0x4L: (Comment missing)
```
why are there 2 0x4L?

cccclai replied:
Remove extra 0x4L in last commit

raziel commented on caffe2/serialize/versions.h:
```diff
@@ -67,7 +67,11 @@ constexpr uint64_t kProducedFileFormatVersion = 0x3L;
+//  0x5L: (update) Update tensor storage schema. For bytecode archive, the
+//  root key of tensor storage is updated from {index} to {the_pointer_value_the_tensor.storage}.
+//  for example: `140245072983168.storage`
+//  Forward-compatibility change. Model size reduced by up to 50%.
```
"Model size reduced by up to 50%." this is hard to tell for sure since it's model dependent (per the analysis you made).
I would remove it.

cccclai replied:
Remove in last commit

raziel reviewed (approved):
Please see my comments - just minor changes.

Thanks

Commit:
Update on "[PyTorch Edge] bytecode version bump to v5 and enable share constant table"


As title, main change:
1. Enable share constant table and reduce model size up to 50%
2. Bump bytecode version from v4 to v5.
3. Add the unittest back. (It was partially removed because `script_module_v5.ptl` bytecode version is v5. When current runtime is v4 and try to load a v5 model, it will raise an error because version is not within the range.

As title

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381/)

Differential Revision: [D28309381](https://our.internmc.facebook.com/intern/diff/D28309381)

[ghstack-poisoned]
